src/SDKs/Azure.Core.Identity/data-plane/Azure.Core.Identity/IdentityClient.cs

```diff
@@ -0,0 +1,155 @@
+                    ClientRequestIdPolicy.Singleton,
+                    options.LoggingPolicy,
+                    BufferResponsePolicy.Singleton);
+        }
+
+        public async Task<AuthenticationResponse> AuthenticateAsync(string tenantId, string clientId, string clientSecret, IEnumerable<string> scopes, CancellationToken cancellationToken = default)
+        {
+            Uri authUri = new Uri(_options.AuthorityHost + tenantId + "/oauth2/token");
+
+            using (var request = _pipeline.CreateRequest())
+            {
+                var body = new Dictionary<string, string>() {
+                    { "response_type", "token" },
+                    { "grant_type", "client_credentials" },
+                    { "client_id", clientId },
+                    { "client_secret", clientSecret },
+                    { "scopes", string.Join(" ", scopes) }
+                };
+
+                request.SetRequestLine(HttpPipelineMethod.Put, authUri);
+
+                ReadOnlyMemory<byte> content = Serialize(body);
+
+                request.Content = HttpPipelineRequestContent.Create(content);
+
+                var response = await _pipeline.SendRequestAsync(request, cancellationToken).ConfigureAwait(false);
+
+                if (response.Status == 200 || response.Status == 201)
+                {
+                    var result = await DeserializeAsync(response.ContentStream, cancellationToken).ConfigureAwait(false);
+
+                    return new Response<AuthenticationResponse>(response, result);
+                }
+
+                else throw new RequestFailedException(response);
+            }
+        }
+
+        private ReadOnlyMemory<byte> Serialize(Dictionary<string,string> dict)
+        {
+            var buff = new byte[1024];
+
+            var writer = new FixedSizedBufferWriter(buff);
+
+            var json = new Utf8JsonWriter(writer);
+
+            json.WriteStartObject();
+
+            foreach(var prop in dict)
+            {
+                json.WriteString(prop.Key, prop.Value);
+            }
+
+            json.WriteEndObject();
+
+            return buff.AsMemory(0, (int)json.BytesWritten);
+        }
+
+        private async Task<AuthenticationResponse> DeserializeAsync(Stream content, CancellationToken cancellationToken)
+        {
+            using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cancellationToken).ConfigureAwait(false))
+            {
+                var response = new Dictionary<string, string>();
+
+                foreach(var property in json.RootElement.EnumerateObject())
+                {
+                    response.Add(property.Name, property.Value.GetString());
+                }
+
+                return new AuthenticationResponse(response);
+            }
+        }
+
+        // TODO (pri 3): CoreFx will soon have a type like this. We should remove this one then.
+        internal class FixedSizedBufferWriter : IBufferWriter<byte>
```

Comment:
@ahsonkhan, do you guys already have a build in byte[] writer?

Reply:
Yes, `ArrayBufferWriter<byte>` in System.Memory.dll, **but that is only available on .NET Core 3.0**.

https://github.com/dotnet/corefx/blob/648d0724c2b9224cace4b1e179fc20904fe27aa4/src/System.Memory/ref/System.Memory.cs#L156-L169

https://github.com/dotnet/corefx/blob/648d0724c2b9224cace4b1e179fc20904fe27aa4/src/Common/src/System/Buffers/ArrayBufferWriter.cs#L17

Reply:
Hmmm, I assumed it would be in the JSON package. @terrajobst

Reply:
Based on previous discussions, I placed the type in an assembly where it would "naturally" live, i.e. where the `IBufferWriter` interface lives. It seems odd to have a general purpose re-sizeable output sink live in a JSON assembly.

Is there significant value here to contort the location of types like this?